router/src/dml_handlers/retention_validator.rs

```diff
@@ -0,0 +1,131 @@
+use std::{ops::DerefMut, sync::Arc};
+
+use async_trait::async_trait;
+use data_types::{DeletePredicate, NamespaceId, NamespaceName, TableId};
+use hashbrown::HashMap;
+use iox_catalog::interface::{get_schema_by_name, Catalog};
+use iox_time::{SystemProvider, TimeProvider};
+use mutable_batch::MutableBatch;
+use observability_deps::tracing::*;
+use thiserror::Error;
+use trace::ctx::SpanContext;
+
+use super::DmlHandler;
+use crate::namespace_cache::{metrics::InstrumentedCache, MemoryNamespaceCache, NamespaceCache};
+
+/// Errors emitted during retention validation.
+#[derive(Debug, Error)]
+pub enum RetentionError {
+    /// The requested namespace could not be found in the catalog.
+    #[error("failed to read namespace schema from catalog: {0}")]
+    NamespaceLookup(iox_catalog::interface::Error),
+
+    /// Time is outside the retention period.
+    #[error("the write is outside of the retention period")]
+    OutsideRetention,
```

Comment:
Maybe we add the name to help the user discover which table contains the data outside the retention window:

```suggestion
    #[error("data in table {0} is outside of the retention period")]
    OutsideRetention(String),
```